From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Fri, 10 Jun 2022 16:51:19 +0500
Subject: [PATCH] feat: entity tracker check Y

Helps to hide minecarts in mineshafts and reduce the server load

diff --git a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
index 96e7e260fcb8eef041bb7ae423eb5fc16d321f2f..a421c6d59e77cda090a7533fb3fcfd2d2e4c0b85 100644
--- a/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
+++ b/src/main/java/com/github/ruviolence/reaper/ReaperConfig.java
@@ -259,4 +259,9 @@ public class ReaperConfig {
     private static void pingPeriod() {
         pingPeriod = MathHelper.clamp(getInt("ping-period", 15), 1, 30) * 1000L;
     }
+
+    public static boolean entityTrackerCheckY;
+    private static void entityTrackerCheckY() {
+        entityTrackerCheckY = getBoolean("entity-tracker-check-y", false);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index 7bcac8e5a0a40275313b11ec0bad359857887819..e286b56ce2d942b0dbc970317fb80cb652bb1b06 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -511,6 +511,13 @@ public class EntityTrackerEntry {
         // Reaper end
         int i = Math.min(this.e, entityplayer.getFurthestViewableBlock()); // Reaper - Dynamic view distance
 
+        // Reaper start
+        if (com.github.ruviolence.reaper.ReaperConfig.entityTrackerCheckY) {
+            double y0 = entityplayer.locY - this.tracker.locY;
+            if (y0 < -i || y0 > i) {
+                return false;
+            }
+        } // Reaper end
         return d0 >= (double) (-i) && d0 <= (double) i && d1 >= (double) (-i) && d1 <= (double) i && this.tracker.a(entityplayer)
                 && (!com.github.ruviolence.reaper.ReaperConfig.delayedChunksEnabled || !entityplayer.getDelayedChunks().contains(new ChunkCoordIntPair(this.tracker.ab, this.tracker.ad))); // Reaper - Delayed chunks
     }
